Backend features - initial setup

maydeff opened the pull request.

maydeff committed:
Initialized EF Core and Autofac to the project, added models, introduced dependency injection

johnnythemadman commented on server/SmartHome.API/Startup.cs:
```diff
@@ -18,6 +26,8 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddDbContext<SmartHomeContext>(options => options.UseSqlServer(ConfigSet.ConnectionString));
```
@Snooking take a look

maydeff commented on server/SmartHome.API/Startup.cs:
```diff
@@ -18,6 +26,8 @@ public Startup(IConfiguration configuration)
         {
+            services.AddDbContext<SmartHomeContext>(options => options.UseSqlServer(ConfigSet.ConnectionString));
+            //TODO: Need mvc, without it nothing will work, need to delete this.
             services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
```
@Snooking take a look

Snooking commented on server/SmartHome.API/Startup.cs:
```diff
@@ -6,25 +6,33 @@
 {
     public class Startup
     {
+        private string WorkAs => "Mayday";
```
No xD

maydeff replied:
XD mark as defect

Snooking commented on server/SmartHome.API/appsettings.Development.json:
```diff
@@ -7,6 +7,7 @@
   },
   "ConnectionStrings": {
-    "SmartHomeDatabase": "Server=.\\SQLEXPRESS;Initial Catalog=SmartHome;Integrated Security=True;"
+    "SmartHomeMayday": "Server=DESKTOP-CHIUD2C;Initial Catalog=SmartHome;Integrated Security=True;",
```
Only one connection string, add this file to git ignore

Snooking replied:
Actually, You should change the name of your server, since it's not a good convention to call it as your Desktop.

Snooking commented on server/SmartHome.BusinessLogic/SmartHome.BusinessLogic.csproj:
```diff
@@ -5,23 +5,18 @@
-    <Folder Include="Home\Commands\" />
     <Folder Include="Home\CommandHandlers\" />
+    <Folder Include="Home\Commands\" />
   </ItemGroup>
```
I think this ain't needed, You can delete this

Snooking commented on server/SmartHome.API/appsettings.Development.json:
```diff
@@ -5,5 +5,8 @@
     }
+  },
+  "ConnectionStrings": {
+    "SmartHome": "Server=.\\SQLEXPRESS;Initial Catalog=SmartHome;Integrated Security=True;"
```
Change your name of the server, since it's not a good convention to call it as your desktop (this way, You'll have conflicts with anyone)

Snooking commented on server/SmartHome.BusinessLogic/SmartHome.BusinessLogic.csproj:
```diff
@@ -1,12 +1,22 @@
-    <Folder Include="Home\" />
+    <Folder Include="Home\CommandHandlers\" />
+    <Folder Include="Home\Commands\" />
+  </ItemGroup>
```
Remove this, as it's not needed

Snooking commented on server/SmartHome.Data/Infrastructure/Enums/ComponentStateEnum.cs:
```diff
@@ -0,0 +1,10 @@
+        ON,
+        OFF,
+        COLLECTING,
+        UNKNOWN
```
It's not proper way of naming enums in C#
![image](https://user-images.githubusercontent.com/24529099/68515289-bbea8600-0280-11ea-9d27-678793f5b664.png)

Snooking commented on server/SmartHome.Data/Infrastructure/Enums/ComponentTypeEnum.cs:
```diff
@@ -0,0 +1,11 @@
+        LIGHT_SENSOR,
+        HUMIDITY_SENSOR,
+        LIGHT_BULB,
+        SWITCH
```
It's not a proper way of naming enums in C#
![image](https://user-images.githubusercontent.com/24529099/68515314-cb69cf00-0280-11ea-87a7-6fe48ad9d0fa.png)

Snooking commented on server/SmartHome.Data/Infrastructure/Extensions/ToStringExtendedExtension.cs:
```diff
@@ -0,0 +1,44 @@
+                default:
+                    return "Unknown";
+            }
+        }
```
Once You resolve enum namings, You can create `PascalToSentenceConverter` or even not, since it's 100% ok to store this as PascalCase

Snooking commented on server/SmartHome.Data/Migrations/20191108214828_FixedInitialMigration.Designer.cs:
```diff
@@ -0,0 +1,232 @@
+{
+    [DbContext(typeof(SmartHomeContext))]
+    [Migration("20191108214828_FixedInitialMigration")]
+    partial class FixedInitialMigration
```
Since it's first migration it should be called `InitializedDatabase` or something like that, not `Fixed...`. You can't fix something that doesn't exist

Snooking commented on server/SmartHome.Data/Migrations/20191108214828_FixedInitialMigration.cs:
```diff
@@ -0,0 +1,230 @@
+
+namespace SmartHome.Data.Migrations
+{
+    public partial class FixedInitialMigration : Migration
```
Since it's first migration it should be called `InitializedDatabase` or something like that, not `Fixed...`. You can't fix something that doesn't exist

Snooking commented on server/SmartHome.Data/Models/Component.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class Component
+    {
+        public Guid ComponentId { get; set; }
```
Why are You duplicating information?
`Component.ComponentId` is not proper naming convention. Change to `Id`

maydeff replied:
Ummm... where am I duplicating information?
I have my convention for Ids and I will follow it.

Snooking commented on server/SmartHome.Data/Models/ComponentData.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class ComponentData
+    {
+        public Guid ComponentDataId { get; set; }
```
Change to `Id`

maydeff replied:
Nope.
As I want to follow naming convention used earlier for IDs (see other model classes) i definitely should not do this.

Snooking commented on server/SmartHome.Data/Models/ComponentType.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class ComponentType
+    {
+        public Guid ComponentTypeId { get; set; }
```
Change to `Id`

maydeff replied:
nope

Snooking commented on server/SmartHome.Data/Models/Module.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class Module
+    {
+        public Guid ModuleId { get; set; }
```
Change to `Id`

maydeff replied:
nope

Snooking commented on server/SmartHome.Data/Models/Room.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class Room
+    {
+        public Guid RoomId { get; set; }
```
Change to `Id`

maydeff replied:
nope

Snooking commented on server/SmartHome.Data/Models/SmartHomeEntity.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class SmartHomeEntity
+    {
+        public Guid SmartHomeEntityId { get; set; }
```
Change to `Id`

maydeff replied:
nope

Snooking commented on server/SmartHome.Data/Models/User.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class User
+    {
+        public Guid UserId { get; set; }
```
Change to `Id`

maydeff replied:
nope

Snooking commented on server/SmartHome.Data/Models/UserSmartHomeEntity.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class UserSmartHomeEntity
+    {
+
```
Not needed new line

Snooking commented on server/SmartHome.Data/SmartHome.Data.csproj:
```diff
@@ -5,7 +5,17 @@
+
+  <ItemGroup>
+    <Reference Include="Microsoft.EntityFrameworkCore.Relational">
+      <HintPath>..\..\..\..\..\..\..\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.entityframeworkcore.relational\2.2.0\lib\netstandard2.0\Microsoft.EntityFrameworkCore.Relational.dll</HintPath>
```
Do something about this path, it'll work only on your machine

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        public SmartHomeContext(DbContextOptions<SmartHomeContext> options) : base(options)
+        {
+        }
+        public DbSet<Component> Components { get; set; }
```
Add a new line before this

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+            OnUserCreated(modelBuilder);
+            OnUserSmartHomeEntityCreated(modelBuilder);
+        }
+        private void OnComponentCreated(ModelBuilder modelBuilder)
```
Add a new line before this

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+            var entity = modelBuilder.Entity<Component>();
+
+            entity
+                .ToTable("Component");
```
Remove 
`                .ToTable("Component");`

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+                .HasOne<Module>(component => component.Module)
+                .WithMany(modules => modules.Components)
+                .HasForeignKey(component => component.ModuleId)
+                .IsRequired();
```
You're using `x => x.` everywhere else, why is here full naming?

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+           
+            var entity = modelBuilder.Entity<ComponentData>();
+            entity
+               .ToTable("ComponentData");
```
Remove

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        {
+            var entity = modelBuilder.Entity<ComponentType>();
+            entity
+               .ToTable("ComponentType");
```
Remove

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        {
+            var entity = modelBuilder.Entity<Module>();
+            entity
+               .ToTable("Module");
```
Remove

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+                .IsRequired();
+
+            entity
+                .HasOne<ComponentType>(component => component.ComponentType)
```
`<ComponentType>` unneeded, please remove here and everywhere below

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        {
+            var entity = modelBuilder.Entity<Room>();
+            entity
+               .ToTable("Room");
```
Remove

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        {
+            var entity = modelBuilder.Entity<SmartHomeEntity>();
+            entity
+               .ToTable("SmartHomeEntity");
```
Remove

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+                .HasMaxLength(15)
+                .IsFixedLength(true)
+                .IsRequired();
+           
```
Unneeded new line

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+        {
+            var entity = modelBuilder.Entity<User>();
+            entity
+               .ToTable("User");
```
Remove

maydeff replied:
nope
![image](https://user-images.githubusercontent.com/44469644/68529286-5dafb880-02fd-11ea-9ca5-ff8ee30fbe2d.png)

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+                .IsUnique();
+
+
+
```
Only one new line between

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+                .Property(x => x.IsAdmin)
+                .IsRequired();
+        }
+        private void OnUserSmartHomeEntityCreated(ModelBuilder modelBuilder)
```
Missing new line before

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+            
+            var entity = modelBuilder.Entity<UserSmartHomeEntity>();
+            entity
+               .ToTable("UserSmartHomeEntity");
```
Remove

maydeff replied:
nope, without this EF uses plural for table names

Snooking commented on server/SmartHome.Data/SmartHomeContext.cs:
```diff
@@ -0,0 +1,244 @@
+               .WithMany(x => x.UserSmartHomeEntities)
+               .HasForeignKey(x => x.SmartHomeEntityId);
+        }
+
```
Unneeded new line

maydeff committed:
Added appsettings to .gitignore.
Cleaned new lines, usings etc.

Made SmartHome.Data to use core 2.2 sdk.
Deleted reference to local dll and replaced it with nuget.

Cleaned up context class.
Moved IsAdmin flag to UserSmartHomeEntity.

Changed enum values to hit naming convention.

maydeff committed:
Added updated InitialMigration.

maydeff committed:
Updated InitialMigration with filled up Up and Down methods.
